neqo-transport/src/frame.rs

```diff
@@ -263,43 +263,71 @@ impl Frame {
         fin: bool,
         space: usize,
     ) -> Option<(Frame, usize)> {
-        let mut remaining = space.saturating_sub(1 + Encoder::varint_len(stream_id));
+        let mut overhead = 1 + Encoder::varint_len(stream_id);
         if offset > 0 {
-            remaining = remaining.saturating_sub(Encoder::varint_len(offset));
+            overhead += Encoder::varint_len(offset);
+        }
+
+        if overhead > space {
+            qdebug!(
+                "Frame::new_stream -> None; hdrlen {} > space {}",
+                overhead,
+                space
+            );
+            return None;
         }
-        let (fin, fill) = if data.len() > remaining {
-            if remaining == 0 {
-                return None;
-            }
 
+        let (fin, fill) = if data.len() > (space - overhead) {
             // More data than fits, fill the packet and negate |fin|.
             (false, true)
-        } else if data.len() == remaining {
+        } else if data.len() == (space - overhead) {
             // Exact fit, fill the packet, keep |fin|.
             (fin, true)
         } else {
             // Too small, so include a length.
-            let data_len = min(remaining - 1, data.len());
-            remaining -= Encoder::varint_len(u64::try_from(data_len).unwrap());
-            remaining = min(data.len(), remaining);
-            // In case the added length causes this to spill over, check |fin| again.
-            (fin && remaining == data.len(), false)
+            let data_len = min(space - overhead - 1, data.len());
+            overhead += Encoder::varint_len(u64::try_from(data_len).unwrap());
+
+            // If all data isn't going to make it in the frame, don't keep fin.
+            let keep_fin = data.is_empty() || data.len() <= (space - overhead);
+            (fin && keep_fin, false)
         };
+
+        if overhead > space || (overhead == space && !fin) {
```

Comment:
Note that the second test is redundant with the test on line 307: `data_len == 0 && !fin`.  You only need the first check to prevent underflow on the `data_len` calculation.